File: problems/0101-0200/0146_lru-cache/solution.go
package main

type listNode struct {
	key   int
	value int
	prev  *listNode
	next  *listNode
}

type LRUCache struct {
	capacity int
	nodes    map[int]*listNode
	head     *listNode
	tail     *listNode
}

func NewLRUCacheTyped(capacity int) *LRUCache {
	cache := &LRUCache{
		capacity: capacity,
		nodes:    make(map[int]*listNode),
		head:     &listNode{},
		tail:     &listNode{},
	}
	cache.head.next = cache.tail
	cache.tail.prev = cache.head
	return cache
}

func (cache *LRUCache) unlink(node *listNode) {
	node.prev.next = node.next
	node.next.prev = node.prev
}

func (cache *LRUCache) pushFront(node *listNode) {
	node.next = cache.head.next
	node.prev = cache.head
	cache.head.next.prev = node
	cache.head.next = node
}

func (cache *LRUCache) get(key int) int {
	node, exists := cache.nodes[key]
	if !exists {
		return -1
	}
	cache.unlink(node)
	cache.pushFront(node)
	return node.value
}

func (cache *LRUCache) put(key int, value int) {
	if node, exists := cache.nodes[key]; exists {
		node.value = value
		cache.unlink(node)
		cache.pushFront(node)
		return
	}
	node := &listNode{key: key, value: value}
	cache.nodes[key] = node
	cache.pushFront(node)
	if len(cache.nodes) > cache.capacity {
		victim := cache.tail.prev
		cache.unlink(victim)
		delete(cache.nodes, victim.key)
	}
}

File: problems/0101-0200/0146_lru-cache/solution.rs
use std::collections::HashMap;

// Slots live in one arena and never move, so a key's node is reached in O(1)
// through the map and unlinked in O(1) through its neighbours. Sentinels at
// HEAD and TAIL remove every boundary case: the list is never empty, so no
// branch has to ask whether a neighbour exists.
const HEAD: usize = 0;
const TAIL: usize = 1;

struct Slot {
    key: i32,
    value: i32,
    prev: usize,
    next: usize,
}

pub struct LRUCache {
    capacity: usize,
    slots: Vec<Slot>,
    index: HashMap<i32, usize>,
    free: Vec<usize>,
}

impl LRUCache {
    pub fn new(capacity: i32) -> Self {
        let sentinel = |prev, next| Slot {
            key: 0,
            value: 0,
            prev,
            next,
        };
        LRUCache {
            capacity: capacity as usize,
            slots: vec![sentinel(TAIL, TAIL), sentinel(HEAD, HEAD)],
            index: HashMap::new(),
            free: Vec::new(),
        }
    }

    fn unlink(&mut self, slot: usize) {
        let (prev, next) = (self.slots[slot].prev, self.slots[slot].next);
        self.slots[prev].next = next;
        self.slots[next].prev = prev;
    }

    fn push_front(&mut self, slot: usize) {
        let first = self.slots[HEAD].next;
        self.slots[slot].prev = HEAD;
        self.slots[slot].next = first;
        self.slots[first].prev = slot;
        self.slots[HEAD].next = slot;
    }

    pub fn get(&mut self, key: i32) -> i32 {
        match self.index.get(&key).copied() {
            Some(slot) => {
                self.unlink(slot);
                self.push_front(slot);
                self.slots[slot].value
            }
            None => -1,
        }
    }

    pub fn put(&mut self, key: i32, value: i32) {
        if let Some(slot) = self.index.get(&key).copied() {
            self.slots[slot].value = value;
            self.unlink(slot);
            self.push_front(slot);
            return;
        }
        if self.index.len() == self.capacity {
            // The tail's neighbour is the least recently used entry; its slot
            // is recycled rather than grown, so the arena stays at capacity.
            let victim = self.slots[TAIL].prev;
            self.unlink(victim);
            let evicted = self.slots[victim].key;
            self.index.remove(&evicted);
            self.free.push(victim);
        }
        let slot = match self.free.pop() {
            Some(reused) => {
                self.slots[reused].key = key;
                self.slots[reused].value = value;
                reused
            }
            None => {
                self.slots.push(Slot {
                    key,
                    value,
                    prev: HEAD,
                    next: HEAD,
                });
                self.slots.len() - 1
            }
        };
        self.index.insert(key, slot);
        self.push_front(slot);
    }
}
